GUI: Added (partial) BiDi support

ZvikaZ opened the pull request:
Added GNU FriBidi, thus allowing Hebrew (or other future RTL languages)
to be displayed correctly.
It's been implemented for all ScummVM GUI (as far as I have noticed),
and can be further used by the engines as needed.

This work is only partial, because for complete BiDi support we'll need
to mirror the widgets, and support input text areas (which currently
don't even support Hebrew text input at all).

Some changes are required in order to to use this:
- Visual Studio:
  -- add FriBidi lib from https://github.com/ShiftMediaProject/fribidi
     (and place the files in the other libs location)
  -- add fribidi.dll to the current directory
  -- add fribidi.lib to VS library list
     (in GUI it's: Project -> scummvm Properties -> Linker ->
                   input -> Additional Dependencies)
  -- Add USE_FRIBIDI define
         either to:
             dists/msvc/ScummVM_Global.props   <PreprocessorDefinitions>
         or in GUI:
             Project -> scummvm Properties ->
             C/C++ -> Preprocessor -> Preprocessor Definitions
- GCC:
  -- install FriBidi
     e.g.,
        Ubuntu: `apt-get install libfribidi-dev`
        CentOS: `yum install fribidi-devel`
  -- rerun `configure`

<!---
Thank you for contributing to ScummVM. Please read the following carefully before submitting your Pull Request.

Make sure your individual commits follow the guidelines found in the ScummVM Wiki: https://wiki.scummvm.org/index.php?title=Commit_Guidelines. If they're not please edit them before submitting the Pull Request.

Commits and Pull Requests should use the following template:

```
SUBSYSTEM: Short (50 chars or less) summary of changes

More detailed explanatory text, if necessary.  Wrap it to about 72
characters or so.  In some contexts, the first line is treated as the
subject of an email and the rest of the text as the body.  The blank
line separating the summary from the body is critical (unless you omit
the body entirely); tools like rebase can get confused if you run the
two together.

Write your commit message in the present tense: "Fix bug" and not "Fixed
bug."  This convention matches up with commit messages generated by
commands like git merge and git revert.

Further paragraphs come after blank lines.

- Bullet points are okay, too

- Typically a hyphen or asterisk is used for the bullet, preceded by a
 single space, with blank lines in between, but conventions vary here

- Use a hanging indent
```
--->

criezy commented on gui/ThemeEngine.cpp:
```diff
@@ -919,7 +919,7 @@ void ThemeEngine::drawDDText(TextData type, TextColor color, const Common::Rect
 
 	_vectorRenderer->setFgColor(_textColors[color]->r, _textColors[color]->g, _textColors[color]->b);
-	_vectorRenderer->drawString(_texts[type]->_fontPtr, text, area, alignH, alignV, deltax, ellipsis, dirty);
+	_vectorRenderer->drawString(_texts[type]->_fontPtr, TransMan.bidiAlgo(text), area, alignH, alignV, deltax, ellipsis, dirty);
```
This will break the compilation when translations is disabled (i.e. if `USE_TRANSLATION` is not defined).

ZvikaZ replied:
Done.
I've also updated the initial PR description a little to much the updated compilation instructions.

sev- commented on snapcraft.yaml:
```diff
@@ -39,6 +39,7 @@ parts:
       - libfreetype6-dev
       - zlib1g-dev
       - libunity-dev
+	  - libfribidi-dev
```
This has extra spaces.

ZvikaZ replied:
Done

sev- commented on common/translation.cpp:
```diff
@@ -450,6 +454,55 @@ bool TranslationManager::checkHeader(File &in) {
+}
+#else
+String TranslationManager::bidiAlgo(const String input) {
+	return input;
```
Does it compile or it says that it loses `const` qualifier?

ZvikaZ replied:
It compiled, both in Windows 10 VS, and Ubuntu.

SupSuper replied:
You should be using `const String &` instead of `const String` in function parameters. This avoids needlessly copying immutable strings.

ZvikaZ replied:
@SupSuper , done

sev- commented on common/translation.cpp:
```diff
@@ -450,6 +454,55 @@ bool TranslationManager::checkHeader(File &in) {
 }
 
+#ifdef USE_FRIBIDI
+String TranslationManager::bidiAlgo(const String input) {
```
I am not a big fan of this name. Perhaps `convertBiDiString` could be better?

ZvikaZ replied:
Done

sev- commented on common/translation.cpp:
```diff
@@ -450,6 +454,55 @@ bool TranslationManager::checkHeader(File &in) {
+		return input;
+
+	if (getCurrentCharset() != "iso-8859-8") {
+		warning("Unexpected charset is used with %s language: %s", getCurrentLanguage().c_str(), getCurrentCharset().c_str());
```
It is better to put name of the method as prefix to the warning, so it would be easier to find it.

ZvikaZ replied:
Done

ZvikaZ commented:
I will fix your notes tomorrow.

Meanwhile, I didn't understand your first question.
What's (non) AA fonts?

criezy commented:
@ZvikaZ It looks better now. Thank you. I will look at this in details in a few days.

@sev- I think it might find the font in `dists/engine-data/fonts.dat`.

digitall commented:
@ZvikaZ : AA in this context is Anti-Aliasing i.e. https://en.wikipedia.org/wiki/Font_rasterization and https://en.wikipedia.org/wiki/Spatial_anti-aliasing

ZvikaZ commented:
OK, but I still don't understand the potential risk with non AA fonts.

And how can I test that?
I don't remember seeing a setting for AA or non AA fonts...

ZvikaZ commented:
Implemented all the suggestions.
Thanks.

Also, I've changed the API a little bit - in the previous version, it assumed that the text is in the GUI language, but game's language can be different than the GUI's.

Regarding the non AA fonts - I tried running with `GUI renderer` equals `Antialiased renderer` and `Standard rendered`, in Hebrew and in English, and didn't see any change.
Anyway, as I said before, I don't see why it should make any difference.

sev- commented:
The difference for the non-AA renderer is fonts. Where does the Hebrew font come from?

ZvikaZ commented:
Honestly, I don't know.
But it's out if the scope of this PR.
Hebrew was displayed before, it was just (almost) mirrored.
This PR adds code manipulates the string to display - it takes an input string, and changes the order of the characters inside - returning another string.

ZvikaZ commented:
I have re-read my initial comment here, and maybe I wasn't clear enough.
When I spoke about 'display correctly' I didn't mean in the sense  of render correctly, but in the sense of displayed in the correct order.

criezy commented:
Also I confirmed that it is getting the font from our `font.dat` file. If I rename that file I can no longer use the Hebrew translation (neither with the AA renderer, nor with the non-AA renderer). But if the `font.dat` file is here, then it works (both with the AA renderer and with the non-AA renderer).

ZvikaZ commented:
> I have one question: if we build without fribidi, do we want to still show Hebrew in the language list in the options? I am guessing it is probably hard to read when all the text is backward. So maybe it should be skipped in the `TranslationManager::getSupportedLanguageNames()` implementation in such a case?

That's a good question...
I took some time to think about it.
Personally, I'm annoyed by the current backward Hebrew.
However, maybe there are users that speak only Hebrew, or feel much more comfortable in Hebrew than English, that they would prefer backward Hebrew to anything else.
Therefore, I think that it's OK to leave Hebrew in even if bidi is disabled, and let the user decide whether he wants that or not.

sev- commented:
Great, thank you, merging.

BLooperZ commented:
Regarding that last question, I suggest allowing users to choose they prefered language for ScummVM GUI.

ZvikaZ commented:
> Regarding that last question, I suggest allowing users to choose they prefered language for ScummVM GUI.

Of course the user can choose.
The question was whether we want to insert Hebrew as one of the options, in platforms that won't support this new BiDi PR.

i.e.,
In most platforms, there will BiDI support, and Hebrew will be displayed correctly, and can be chosen.
But there might be platforms that GNU's FriBidi library isn't available for them, and Hebrew will be displayed reversed. 
So - the question is, do we want to leave Hebrew as an option in that case.
The current decision is yes - leave Hebrew as an option, even if it's reversed, and let the user decide.

antoniou79 commented:
Wouldn't it be possible to integrate the FriBiDi source code in ScummVM so as to avoid the third party library dependency (and the cross-compilation it will require per port)?

For example, there's currently work done for Hebrew subtitles for Blade Runner by rzil. Ideally, I'd like to make use of the BiDi support in ScummVM without the extra implication that some of our ports will not be displaying the subtitles correctly in Hebrew.

ZvikaZ commented:
> Wouldn't it be possible to integrate the FriBiDi source code in ScummVM so as to avoid the third party library dependency (and the cross-compilation it will require per port)?
> 
> For example, there's currently work done for Hebrew subtitles for Blade Runner by rzil. Ideally, I'd like to make use of the BiDi support in ScummVM without the extra implication that some of our ports will not be displaying the subtitles correctly in Hebrew.

Well, that's a question of policy, I think.
After all, why didn't we integrate other libs inside ScummVM?

@sev- when do we prefer to have a library as an external dependancy, and when do we integrate it as part of ScummVM?

And as for rzil (@BLooperZ ) work, we've been in touch in the recent days, and we're coordinated. As far as I understand, he's happy that we have now a good BiDi library available.

lephilousophe commented:
IMO it would be a bad idea to include Fribidi in the main tree as it's an external dependency receiving bugfixes on a regular basis. Including it would make it necessary to follow upstream.
In addition, as @ZvikaZ said, why don't we include other libraries too?
The only library I see which is included in the tree is the Lua one and it seems we patched it.

I already added fribidi to almost all platforms supported by [dockerized-bb](https://github.com/lephilousophe/dockerized-bb/). Builds are still compiling but it seems OK so far.
It was really easy to add the dependency.

antoniou79 commented:
Hmm, I didn't think FriBiDi was receiving regular updates.

It seems to me that most of the other third party libraries that we don't include are either optional or affecting specific (small or just one) other engine(s). 

I am also unsure about their dependencies, meaning that in the case of FriBiDi they state that it does not depend on any other library, so it would be straight forward to integrate. I don't think this is the case for some of our other third party libs.

lotharsm commented:
Maybe we should move this out of this PR since this could lead to a more "general" discussion - scummvm-devel@ might be a better place.

Have we ever bundled third-party libs in the past despite `munt`? I'm not a big fan of this idea since I think this will introduce a pretty high workload due to having to keep track about updates.
